src/clarity-angular/tree-view/abstract-tree-selection.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2016 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+/**
+ * Author: Eudes
+ */
+
+import {TreeSelection} from "./tree-selection";
+import {TreeSelectionService} from "./providers/treeSelection.service";
+
+export abstract class AbstractTreeSelection {
+    constructor(public parent: AbstractTreeSelection,
+                public treeSelectionService: TreeSelectionService) {
+    }
+
+    public model: any;
+
+    abstract get children(): AbstractTreeSelection[];
+
+    abstract selectedChanged(): void;
+
+    private _selected: boolean = false;
+    private _indeterminate: boolean = false;
+
+    public get selected(): boolean {
+        return this._selected;
+    }
+
+    public set selected(value: boolean) {
+        this._selected = value;
+        this._indeterminate = false;
+        this.children.forEach(child => child.parentChanged(value));
+        if (this.parent) {
+            this.parent.childChanged();
+        }
+        this.treeSelectionService.notify();
+        this.selectedChanged();
+    }
+
+    public get indeterminate(): boolean {
+        return this._indeterminate;
+    }
+
+    childChanged(): void {
+        let oneSelectedChild = false;
+        let previousValue: boolean = this._selected;
+        this._selected = true;
+        this._indeterminate = false;
+
+        for (let child of this.children) {
+            if (child.indeterminate) {
+                this._selected = false;
+                this._indeterminate = true;
+                break;
+            }
+            if (child.selected) {
+                oneSelectedChild = true;
+                if (this._selected === false) {
+                    this._indeterminate = true;
+                    break;
+                }
+            } else {
+                this._selected = false;
+                if (oneSelectedChild) {
+                    this._indeterminate = true;
+                    break;
+                }
+            }
+        }
+        // FIXME: Only update the parent if this node really changed
+        // But when doing that, we need to make sure the root node is still warned that the overall selection
+        // has changed.
+        if (this.parent) {
+            this.parent.childChanged();
+        }
```

Comment:
You can handle that FIXME right now: by having the TreeView observe the singleton service instead of waiting for its children to bubble up a change, you can stop the propagation from child to parent as soon as one of the parents is unchanged.